**Завдання 1. Оптимальна ціна товару**

Маємо:\
Функція попиту на навушники:\
$Q(x)=2000-0.8x$, де $x$-ціна в гривнях, $Q(x)$-кількість проданих одиниць на місяць.\
Функція прибутку:\
$P(x)=(x-800)(2000-0.8x)$\
Собівартість одних навушників становить 800 грн.\
Тоді прибуток компанії:
$$P(x)=(x-800)⋅ Q(x)=(x-800)(2000-0.8x)$$
- - -
Рішення:
1. Знайдемо критичну точку:\
Розкриємо функцію прибутку:
$$P(x)=(x-800)(2000-0.8x)=2000x-0.8^2-1600000+640x=-0.8x^2+2640x-1600000$$
Знайдемо похідну:
$$P'(x)=-1.6x+2640$$
Прирівняємо до нуля $0$ та вирішим рівняння:
$$-1.6х+2640=0 $$
$$x=1650$$
Критична точка: $x=1650\ грн$

2. Тип екстремуму:\
Друга похідна:$$P''(x)=-1.6<0$$
Оскільки $P''(x)<0$, то $x=1650$ є точкою максимуму.


In [1]:
#3. Чисельна оптимізація
from scipy.optimize import minimize_scalar

# Функція прибутку (мінімізуємо -P(x) для максимізації P(x))
def profit(x):
    return -((x - 800) * (2000 - 0.8 * x))

# Пошук оптимуму на інтервалі [800, 2500]
result = minimize_scalar(profit, bounds=(800, 2500), method='bounded')

optimal_price = result.x
max_profit = -result.fun

print(f"Оптимальна ціна: {optimal_price:.2f} грн")
print(f"Максимальний прибуток: {max_profit:.2f} грн")


Оптимальна ціна: 1650.00 грн
Максимальний прибуток: 578000.00 грн


4. Порівняння аналітичного та чисельного методу оптимізації:\
При оптимальній ціні $1650\ грн$:
*   Кількість проданих одиниць:
$$ Q(1650)=2000-0.8⋅1650=680\ шт$$
*   Максимальний прибуток:
$$ P(1650)=(1650-800)⋅680=578000\ грн$$

**Висновок:**
Компанія максимізує прибуток при ціні $1650\ грн$, отримавши $578000\ грн$ місячного прибутку.
- - -

**Завдання 2. Мінімізація функції двох змінних**

Маємо функцію:
$f(x, y)=x^2+xy+y^2-6x-9y+20$
1. Пошук критичної точки:
Знайдемо частинні похідні функції $f(x, y)=x^2+xy+y^2-6x-9y+20$
$$\frac {\partial f}{\partial x}=2x+y-6$$
$$\frac {\partial f}{\partial y}=x+2y-9$$
Прирівняємо їх до нуля та розв'яжемо систему лінійних рівнянь:
\begin{equation}
\begin{cases}
2x+y-6 = 0 \\
x+2y-9=0
\end{cases}
\Rightarrow
\begin{cases}
2x+y=6\\
x+2y=9
\end{cases}
\end{equation}
Виразимо $x$ з другого рівняння:
$$ x=9-2y$$
Підставимо в перше рівняння:
$$2(9-2y)+y=6 $$
$$18-4y+y=6$$
$$-3y=-12 \Rightarrow y=4$$
Знайдемо $x$:
$$x=9-2(4)=1$$
**Критична точка: $(1; 4)$**

2. Перевірка типу екстремуму (Матриця Гессе)
Знайдемо другі частинні похідні:
$$ f''_{xx}=2$$
$$ f''_{xy}=1$$
$$ f''_{yx}=1$$
$$ f''_{yy}=2$$
Складемо матрицю Гессе $Н$:
$$ H=\left( \begin{matrix} 2 & 1 \\ 1 & 2 \end {matrix} \right) $$
Складемо хаарактеристичне рівняння $det(H-λI)=0$:
$$ det\left( \begin{matrix} 2-λ & 1 \\ 1 & 2-λ \end {matrix} \right)=(2-λ)(2-λ)-1=0 $$
$$ (2-λ)^2=1 $$
$$ (2-λ)=\pm1$$
Звідси отримуємо власні значення:
$$2-λ_1=1 \Longrightarrow  λ_1=1$$
$$2-λ_2=-1 \Longrightarrow  λ_2=3$$
**Висновок:**\
Оскільки обидва значення додатні, матриця Гессе додатно визначена, отже, знайдена точка (1;4) є точкою локального мінімуму.


In [10]:
# 3. Знайдемо оптимум чисельно.
import numpy as np
from scipy.optimize import minimize

# Визначення функції
def f(args):
    x, y = args
    return x**2 + x*y + y**2 - 6*x - 9*y + 20

# Початкові точки для перевірки стійкості
starting_points = [(0, 0), (10, 10), (-5, 15)]

print(f"{'Start Point':<15} | {'Min Point (x, y)':<17} | {'Function Value':<15}")
print("-" * 52)

for start in starting_points:
    # Запуск оптимізації методом BFGS
    result = minimize(f, start, method='BFGS')

    # Форматування результату
    min_x, min_y = result.x
    print(f"{str(start):<15} | ({min_x:.4f}, {min_y:.4f})  | {result.fun:.4f}")

    # Перевірка збіжності
    if not result.success:
        print(f"Optimization failed for start point {start}")


Start Point     | Min Point (x, y)  | Function Value 
----------------------------------------------------
(0, 0)          | (1.0000, 4.0000)  | -1.0000
(10, 10)        | (1.0000, 4.0000)  | -1.0000
(-5, 15)        | (1.0000, 4.0000)  | -1.0000


**Висновок:** Усі три запуску збігаються до  однієї точки $(1,4)$, що підтверджує стійкість розв’язку та правильність знаходження глобального мінімуму.
- - -

**Завдання 3. Оптимальний план виробництва**\
Меблева фабрика виготовляє два види продукції: стільці та столи. Для виробництва використовуються два ресурси: деревина та робочий час.  
Ресурси та прибуток:
*   Деревина: $2 м^2$ на стілець,  $4 м^2$ на стіл, загальний запас - $120 м^2$
*   Робочий час: 3 години на стілець, 2 години на стіл, запас часу - 90 годин
*   Прибуток: 500 грн за стілець, 800 грн за стіл.\
Потрібно знайти оптимальний план виробництва, який максимізує прибуток.
- - -
Рішення:\
1.Сформулюємо задачу лінійного програмування аналітично:\
Позначимо:
*   $x$-кількість стільців
*   $y$-кількість столів\
Цільова функція прибутку:\
$$P(x, y)=500x+800y \to max$$
Обмеження:
*   Деревина: $2x+4y≤120$
*   Робочий час: $3x+2y≤90$
*   Невід’ємність: $x\ge0$, $y\ge0$







In [1]:
# Програмний розв'язок завдання
import numpy as np
from scipy.optimize import linprog

# Коефіцієнти цільової функції (мінімізуємо -P)
c = [-500, -800]

# Матриця коефіцієнтів обмежень (A_ub @ x <= b_ub)
A_ub = [
    [2, 4],   # обмеження на деревину
    [3, 2],   # обмеження на робочий час
]

# Праві частини обмежень
b_ub = [120, 90]

# Межі змінних (x >= 0, y >= 0)
bounds = [(0, None), (0, None)]

# Розв'язання задачі
res = linprog(c=c, A_ub=A_ub, b_ub=b_ub, bounds=bounds, method='highs')

# Виведення результатів
x_opt, y_opt = res.x
max_profit = -res.fun

print(f"Статус: {res.message}")
print(f"Оптимальний план: x = {x_opt:.2f} стільців, y = {y_opt:.2f} столів")
print(f"Максимальний прибуток: {max_profit:.2f} грн")

# Обчислення використання ресурсів
wood_used = 2*x_opt + 4*y_opt
time_used = 3*x_opt + 2*y_opt
wood_left = 120

Статус: Optimization terminated successfully. (HiGHS Status 7: Optimal)
Оптимальний план: x = 15.00 стільців, y = 22.50 столів
Максимальний прибуток: 25500.00 грн


Отриманий оптимальні значення:

*  Стільців $(x)=15.0$; столів $(y)=22.5$
*  Максимальний прибуток: 25500 грн

Аналіз ресурсів:
$$Деревина:  2x+4y \Rightarrow 2⋅15+4⋅22.5=30+90=120 $$
Використано деревини: 120.0 м², залишок 0.0 м².
$$Робочий\ час:  3x+2y=3⋅15+2⋅22.5=45+45=90$$
Використано робочого часу: 90.0 год, залишок 0.0 год.



In [10]:
# Завдання 4. Комплексний аналіз виробництва
import numpy as np
from scipy.optimize import approx_fprime, minimize_scalar, minimize
from scipy.integrate import quad

# Частина 1: Дослідження продуктивності
def productivity(t):
    return 100 + 40 * t - 4 * t**2
def productivity_for_gradient(t_array):
    t = t_array[0]
    return productivity(t)

print("Частина 1: Дослідження продуктивності")

# 1. Аналіз динаміки (похідна) в точках t = 2, 5, 8
time_points = [2, 5, 8]
epsilon = 1e-6
print("1. Швидкість зміни продуктивності P'(t):")
for t in time_points:
    deriv = approx_fprime([t], productivity_for_gradient, epsilon)[0]
    state = "зростає" if deriv > 0 else "спадає" if deriv < 0 else "пік/стабільно"
    print(f"   t = {t} год: P'(t) ≈ {deriv:.2f} -> Продуктивність {state}")

# 2. Пошук моменту пікової продуктивності
res_max = minimize_scalar(lambda t: -productivity(t), bounds=(0, 10), method='bounded')
t_peak = res_max.x
p_peak = productivity(t_peak)

print(f"\n2. Пікова продуктивність:")
print(f"   Час (t*): {t_peak:.2f} год")
print(f"   Максимальне значення P(t*): {p_peak:.2f} од/год")

# 3. Загальний обсяг виробництва (інтеграл)
total_production, error = quad(productivity, 0, 10)
print(f"\n3. Загальний обсяг виробництва за зміну:")
print(f"   Інтеграл P(t) від 0 до 10: {total_production:.2f} одиниць")

# Частина 2: Оптимізація витрат

print("\nЧастина 2: Оптимізація витрат")

# 1. Визначення початкових параметрів (x0, y0) через систему рівнянь
# Система:
# 2x + y = 20
# x + 3y = 25

A = np.array([
    [2, 1],
    [1, 3]
])
B = np.array([20, 25])

initial_params = np.linalg.solve(A, B)
x0, y0 = initial_params

print(f"1. Початкові параметри (з бюджетних обмежень):")
print(f"   x0 (сировина) = {x0:.2f}")
print(f"   y0 (енергія)  = {y0:.2f}")

# 2. Мінімізація вартості виробництва
def cost_function(vars):
    x, y = vars
    return x**2 + y**2 - 10*x - 8*y + 50

# Використовуємо початкову точку (x0, y0) з попереднього кроку
res_min_cost = minimize(cost_function, initial_params, method='BFGS')

x_opt, y_opt = res_min_cost.x
min_cost_value = res_min_cost.fun

print(f"\n2. Оптимізація витрат (метод BFGS):")
print(f"   Оптимальні витрати на сировину (x*): {x_opt:.2f}")
print(f"   Оптимальні витрати на енергію (y*): {y_opt:.2f}")
print(f"   Мінімальна вартість одиниці C(x*, y*): {min_cost_value:.2f} грн")


# Фінал
print("\n Фінал")
total_cost_final = total_production * min_cost_value

print(f"Загальна вартість виробництва всієї партії:")
print(f"{total_production:.2f} (обсяг) * {min_cost_value:.2f} (вартість/од) = {total_cost_final:.2f} грн")

Частина 1: Дослідження продуктивності
1. Швидкість зміни продуктивності P'(t):
   t = 2 год: P'(t) ≈ 24.00 -> Продуктивність зростає
   t = 5 год: P'(t) ≈ -0.00 -> Продуктивність спадає
   t = 8 год: P'(t) ≈ -24.00 -> Продуктивність спадає

2. Пікова продуктивність:
   Час (t*): 5.00 год
   Максимальне значення P(t*): 200.00 од/год

3. Загальний обсяг виробництва за зміну:
   Інтеграл P(t) від 0 до 10: 1666.67 одиниць

Частина 2: Оптимізація витрат
1. Початкові параметри (з бюджетних обмежень):
   x0 (сировина) = 7.00
   y0 (енергія)  = 6.00

2. Оптимізація витрат (метод BFGS):
   Оптимальні витрати на сировину (x*): 5.00
   Оптимальні витрати на енергію (y*): 4.00
   Мінімальна вартість одиниці C(x*, y*): 9.00 грн

 Фінал
Загальна вартість виробництва всієї партії:
1666.67 (обсяг) * 9.00 (вартість/од) = 15000.00 грн
